math/src/fft/fft_inputs.rs

```diff
@@ -0,0 +1,154 @@
+use super::{permute_index, FieldElement, StarkField};
+
+// CONSTANTS
+// ================================================================================================
+const MAX_LOOP: usize = 256;
+
+// FFTINPUTS TRAIT
+// ================================================================================================
+
+/// Defines the interface that must be implemented by the input to fft_in_place method.
+pub trait FftInputs<B: StarkField> {
+    /// Returns the number of elements in this input.
+    fn len(&self) -> usize;
+
+    /// Combines the result of smaller number theoretic transform into a larger NTT.
+    fn butterfly(&mut self, offset: usize, stride: usize);
+
+    /// Combines the result of smaller number theoretic transform multiplied with a
+    /// twiddle factor into a larger NTT.
+    fn butterfly_twiddle(&mut self, twiddle: B, offset: usize, stride: usize);
+
+    /// Swaps the element at index i with the element at index j.
+    fn swap(&mut self, i: usize, j: usize);
+
+    /// Multiplies every element in this input by a series of increment. Specifically:
+    ///
+    /// elem_i = elem_i * offset * increment^i
+    fn shift_by_series(&mut self, offset: B, increment: B);
+
+    /// Multiplies every element in this input by `offset`. Specifically:
+    ///
+    /// elem_i = elem_i * offset
+    fn shift_by(&mut self, offset: B);
+
+    /// Permutes the elements in this input using the permutation defined by the given
+    /// permutation index. The permutation index is a number between 0 and `self.len() - 1`
+    /// that specifies the permutation to apply to the input. The permutation is applied
+    /// in place, so the input is replaced with the result of the permutation. The permutation
+    /// is applied by swapping elements in the input.
+    fn permute(&mut self) {
+        let n = self.len();
+        for i in 0..n {
+            let j = permute_index(n, i);
+            if j > i {
+                self.swap(i, j);
+            }
+        }
+    }
+
+    /// Applies the FFT to this input. The FFT is applied in place, so the input is
+    /// replaced with the result of the FFT. The `twiddles` parameter specifies the
+    /// twiddle factors to use for the FFT. The `twiddles` parameter must be a slice
+    /// of length `self.len() / 2`.
```

Comment:
nit: rustdoc does not render monolithic paragraphs like this very well. It is better to have a short intro sentence (1 or  2 lines at most) and then more detailed description. For example:

```
Applies the FFT to this input.

The FFT is applied in place, so the input is  replaced with the result of the FFT. The
`twiddles` parameter specifies the twiddle factors to use for the FFT.

# Panics
Panics if length of the `twiddles` parameter is not self.len() / 2.
```